fix(client): Fix batch order for middleware/query combo

SevInf opened the pull request:
Root cause of the problem: data loader batches requests in the order
they come in, which in case of `$transaction` call happens to be the
order in which they are specified in the array. However, depending on
middlewares/extensions (and possibly other factors), it is possible that
they will be delivered to the data loader in a different order.
Fortunately, we already have an index of each request in a batch as a
part of request, so we can just sort the requests before sending them to
the engine.

Fix #18276

SevInf committed:
fix(client): Fix batch order for middleware/query combo

Root cause of the problem: data loader batches requests in the order
they come in, which in case of `$transaction` call happens to be the
order in which they are specified in the array. However, depending on
middlewares/extensions (and possibly other factors), it is possible that
they will be deleviered to the data loader in a different order.
Fortunately, we already have an index of each request in a batch as a
part of request, so we can just sort the requests before sending them to
the engine.

Fix #18276

millsp commented on packages/client/src/runtime/DataLoader.ts:
```diff
@@ -65,6 +68,7 @@ export class DataLoader<T = unknown> {
             batch[0].reject(e)
           })
       } else {
+        batch.sort((a, b) => this.options.batchOrder(a.request, b.request))
```
Is this a symptom of something we need to fix that lives more deeply, or is it the only way to fix it? I am curious to know why/how we could not preserve the order.

SevInf replied:
So, what happens is:
- we call `prisma.$queryRaw` and await it. This triggers `PrismaPromise` and it passes through all the middlewares and gets to a query extension.
- extension wraps it into `$transaction` and adds 2 extra queries, 1 before and 1 after the original. 
- `$transaction` triggers all 3 `PrismaPromises`. Middle promise (original one) is already in progress and does not need pass through middleware. 2 new requests that extension added still do. As a result, middle promise reaches data loader before either of the outer ones.
-  Due to `lock` promise we won't flush the batch until all Promises have reached data loader (this part works correctly), but `DataLoader` still sends them to the engine in the order they come in, which means, middle promise is sent first. 

I think sorting is simplest solution for this case - we guaranteed to have all necessary promises by the time we flush the batch, we know what order they should be. Even simpler solution would've been re-implementing batching to bypass `DataLoader`, but that would require much more time.

millsp replied:
I think this is a more general problem and is not limited to extensions. This simple code

```ts
prisma.$use(async (params, next) => {
  await new Promise((r) => setTimeout(r, Math.random() * 1000))
  return next(params)
})

await prisma.$transaction([
  prisma.$queryRawUnsafe('SELECT 1'),
  prisma.$queryRawUnsafe('SELECT 2'),
  prisma.$queryRawUnsafe('SELECT 3'),
])
```

millsp replied:
Another way to fix it would be to use the locks to unlock in the right order, but let's rather remove all of this asap.

SevInf committed:
Fix process.nextTick polyifill for edge client

Promise callback is actually called before next callback in the
microtask queue. When locks are used, this causes batch to be flushed
before all requests waiting on lock have finished.
